# Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing Data Set

In [3]:
# Import customer data set
# Create a string of the path
# Define the path variable
path=r'/Users/douniaelyoussoufi/Achievement 4 '
# Load the ords_prods_merge.pkl file directly using os.path.join
customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

# Data Wrangling and Quality Assurance for Consistency and Cleanliness

In [7]:
# Display the first 5 rows 
customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
# Rename columns: Convert to lowercase, replace spaces with underscores, and fix "surnam" typo
customer.columns = customer.columns.str.lower().str.replace(' ', '_')
customer.rename(columns={'surnam': 'surname'}, inplace=True)

In [11]:
# Check the data types of all columns
customer.dtypes

user_id          int64
first_name      object
surname         object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [13]:
# Convert 'date_joined' to datetime format
customer['date_joined'] = pd.to_datetime(customer['date_joined'])

In [15]:
# List of categorical columns to standardize
categorical_columns = ['first_name', 'surname', 'gender', 'state', 'fam_status']
# Standardize all categorical columns: Remove spaces and convert to lowercase
for col in categorical_columns:
    customer[col] = customer[col].str.strip().str.lower()

In [17]:
# Check for any invalid age or negative values in 'n_dependants'
print(customer[customer['age'] < 0])
print(customer[customer['n_dependants'] < 0])

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []
Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []


In [19]:
# Check for missing values
customer.isnull().sum()

user_id             0
first_name      11259
surname             0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [21]:
# Count occurrences of each surname
surname_counts = customer['surname'].value_counts()
# Check if any surname appears more than once
duplicates_exist = (surname_counts > 1).any()
# Print result
if duplicates_exist:
    print("There are duplicate surnames.")
else:
    print("No duplicate surnames found. The 'first_name' column may not be needed.")

There are duplicate surnames.


In [23]:
# Fill missing values in the 'first_name' column with 'Unknown' to handle null entries
customer['first_name'] = customer['first_name'].fillna('Unknown')

In [25]:
# Check for duplicates
duplicates = customer.duplicated().sum()
print("Number of duplicate rows:", duplicates)

Number of duplicate rows: 0


In [27]:
# Check for outliers in 'age' and 'income'
print("Summary statistics for Age and Income:\n", customer[['age', 'income']].describe())

Summary statistics for Age and Income:
                  age         income
count  206209.000000  206209.000000
mean       49.501646   94632.852548
std        18.480962   42473.786988
min        18.000000   25903.000000
25%        33.000000   59874.000000
50%        49.000000   93547.000000
75%        66.000000  124244.000000
max        81.000000  593901.000000


In [29]:
# Show all rows where income is greater than 124,244
customer[customer['income'] > 124244]

user_id   first_name     surname  gender           state  age  \
0         26711      deborah    esquivel  female        missouri   48   
8         69965       jeremy        vang    male           texas   47   
14          516        peter        hunt    male        colorado   51   
21        37744        wanda       salas  female        arkansas   77   
24        57549        shawn       moore    male         montana   72   
...         ...          ...         ...     ...             ...  ...   
206196   139950       gloria      murray  female        colorado   45   
206197    74598  christopher   velazquez    male       minnesota   52   
206199   179673         adam  villanueva    male         wyoming   77   
206204   168073         lisa        case  female  north carolina   44   
206205    49635       jeremy     robbins    male          hawaii   62   

       date_joined  n_dependants        fam_status  income  
0       2017-01-01             3           married  165665  
8       2017-01-01             1           married  162432  
14      2017-01-01             2           married  146559  
21      2017-01-01             1           married  125977  
24      2017-01-01             0  divorced/widowed  135302  
...            ...           ...               ...     ...  
206196  2020-04-01             2           married  150954  
206197  2020-04-01             0            single  140700  
206199  2020-04-01             0  divorced/widowed  162239  
206204  2020-04-01             1           married  148828  
206205  2020-04-01             3           married  168639  

[51551 rows x 10 columns]

In [31]:
# Sort the entire DataFrame by the 'income' column in descending order and get the last 300 rows
print(customer.sort_values('income', ascending=False)['income'].head(300))

21726     593901
202655    592409
81583     591089
111950    590790
167037    584097
           ...  
175764    345689
182790    344681
56468     344624
22483     344569
84108     344356
Name: income, Length: 300, dtype: int64


### High-Income Values Analysis

After inspecting the dataset, it was observed that there are several users with high-income values, including amounts greater than 300,000, with some reaching up to 593,901. These values seem logical based on the context of the dataset, as they represent a range of high-income earners. 

No data issues were identified with these high values, and they are consistent with the expected income distribution. Therefore, these values are not considered outliers and will be retained in the analysis. Further analysis will focus on understanding the income distribution and its impact on the dataset.

In [34]:
# Display the cleaned DataFrame
customer.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    deborah  esquivel  female    missouri   48  2017-01-01   
1    33890   patricia      hart  female  new mexico   36  2017-01-01   
2    65803    kenneth    farley    male       idaho   35  2017-01-01   
3   125935   michelle     hicks  female        iowa   40  2017-01-01   
4   130797        ann   gilmore  female    maryland   26  2017-01-01   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# Combime Databases

In [37]:
# Load the instacart data (ords_prods_merge_48.pkl)
instacart_data = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_48.pkl'))

In [38]:
# Check the data types of the key columns (assuming 'user_id' is the key column)
print(customer['user_id'].dtype)  # Check the dtype of user_id in customer data
print(instacart_data['user_id'].dtype)  # Check the dtype of user_id in instacart data

int64
int64


In [41]:
# Now, merge the data on 'user_id'
merged_data = pd.merge(instacart_data, customer, on='user_id', how='left')  # Use 'left' to keep all Instacart data

In [42]:
# Check the result
merged_data.head()

order_id  product_id  add_to_cart_order  reordered  user_id eval_set  \
0         2       33120                  1          1   202279    prior   
1         2       28985                  2          1   202279    prior   
2         2        9327                  3          0   202279    prior   
3         2       45918                  4          1   202279    prior   
4         2       30035                  5          0   202279    prior   

   order_number  order_dow  order_hour_of_day  days_since_prior_order  ...  \
0             3          5                  9                     8.0  ...   
1             3          5                  9                     8.0  ...   
2             3          5                  9                     8.0  ...   
3             3          5                  9                     8.0  ...   
4             3          5                  9                     8.0  ...   

    order_frequency_flag  first_name  surname  gender  state  age date_joined  \
0  Non-frequent customer        paul  coleman    male  idaho   57  2020-02-06   
1  Non-frequent customer        paul  coleman    male  idaho   57  2020-02-06   
2  Non-frequent customer        paul  coleman    male  idaho   57  2020-02-06   
3  Non-frequent customer        paul  coleman    male  idaho   57  2020-02-06   
4  Non-frequent customer        paul  coleman    male  idaho   57  2020-02-06   

  n_dependants fam_status  income  
0            3    married   98119  
1            3    married   98119  
2            3    married   98119  
3            3    married   98119  
4            3    married   98119  

[5 rows x 32 columns]

# Export the Dataframe

In [49]:
# Export the merged_data dataframe 
merged_data.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'merged_data_part1.pkl'))